File: idea-plugin/src/main/kotlin/com/dmxiaoshubao/copypathline/idea/core/LineRanges.kt
package com.dmxiaoshubao.copypathline.idea.core

object LineRanges {
    fun collect(text: CharSequence, selections: List<SelectionRange>): List<LineRange> {
        if (selections.isEmpty()) {
            return emptyList()
        }

        val lineMap = LineOffsetMap(text)
        return normalize(selections.map { selectionToRange(lineMap, it) })
    }

    fun format(ranges: List<LineRange>): String =
        ranges.joinToString(",") { range ->
            if (range.start == range.end) "${range.start}" else "${range.start}-${range.end}"
        }

    private fun selectionToRange(lineMap: LineOffsetMap, selection: SelectionRange): LineRange {
        val startOffset = selection.startOffset.coerceAtLeast(0)
        val endOffset = selection.endOffset.coerceAtLeast(0)
        val startLine = lineMap.lineForOffset(startOffset) + 1
        var endLine = lineMap.lineForOffset(endOffset) + 1

        if (selection.hasSelection && endOffset > startOffset && lineMap.isLineStart(endOffset) && endLine > startLine) {
            endLine -= 1
        }

        return LineRange(startLine, maxOf(startLine, endLine))
    }

    private fun normalize(ranges: List<LineRange>): List<LineRange> {
        if (ranges.isEmpty()) {
            return emptyList()
        }

        val sorted = ranges.sortedWith(compareBy(LineRange::start, LineRange::end))
        val merged = mutableListOf<LineRange>()

        for (range in sorted) {
            val previous = merged.lastOrNull()
            if (previous == null) {
                merged += range
                continue
            }

            if (range.start <= previous.end + 1) {
                merged[merged.lastIndex] = LineRange(previous.start, maxOf(previous.end, range.end))
                continue
            }

            merged += range
        }

        return merged
    }

    private class LineOffsetMap(text: CharSequence) {
        private val lineStarts: IntArray = buildList {
            add(0)
            text.forEachIndexed { index, char ->
                if (char == '\n' && index + 1 <= text.length) {
                    add(index + 1)
                }
            }
        }.toIntArray()

        fun lineForOffset(offset: Int): Int {
            val normalized = offset.coerceIn(0, maxLineOffset())
            var low = 0
            var high = lineStarts.lastIndex
            var result = 0

            while (low <= high) {
                val middle = (low + high) ushr 1
                if (lineStarts[middle] <= normalized) {
                    result = middle
                    low = middle + 1
                } else {
                    high = middle - 1
                }
            }

            return result
        }

        fun isLineStart(offset: Int): Boolean = lineStarts.binarySearch(offset) >= 0

        private fun maxLineOffset(): Int = lineStarts.last()
    }
}

File: src/core/lineRanges.ts
export interface SelectionInput {
  readonly startLine: number;
  readonly startCharacter: number;
  readonly endLine: number;
  readonly endCharacter: number;
  readonly isEmpty: boolean;
}

export interface LineRange {
  readonly start: number;
  readonly end: number;
}

export function selectionToLineRange(selection: SelectionInput): LineRange {
  const start = selection.startLine + 1;
  let endLine = selection.endLine;

  if (!selection.isEmpty && selection.endCharacter === 0 && selection.endLine > selection.startLine) {
    endLine -= 1;
  }

  const end = endLine + 1;
  return {
    start,
    end: Math.max(start, end)
  };
}

export function normalizeLineRanges(ranges: readonly LineRange[]): LineRange[] {
  if (ranges.length === 0) {
    return [];
  }

  const sorted = [...ranges].sort((left, right) => {
    if (left.start !== right.start) {
      return left.start - right.start;
    }

    return left.end - right.end;
  });

  const merged: LineRange[] = [];

  for (const range of sorted) {
    const previous = merged.at(-1);

    if (!previous) {
      merged.push({ ...range });
      continue;
    }

    if (range.start <= previous.end + 1) {
      merged[merged.length - 1] = {
        start: previous.start,
        end: Math.max(previous.end, range.end)
      };
      continue;
    }

    merged.push({ ...range });
  }

  return merged;
}

export function collectLineRanges(selections: readonly SelectionInput[]): LineRange[] {
  return normalizeLineRanges(selections.map(selectionToLineRange));
}

export function formatLineRanges(ranges: readonly LineRange[]): string {
  return ranges
    .map((range) => (range.start === range.end ? `${range.start}` : `${range.start}-${range.end}`))
    .join(",");
}
